doc/compiler/control/CheckpointRestoreCoordination.md

```diff
@@ -0,0 +1,434 @@
+portability restrictions on the target CPU. It then calls the top level method
+in the compiler that performs the necessary coordination:
+`TR::CompilationInfo::prepareForRestore()`. Once the coordination is complete,
+the executing thread returns to the VM, which eventually resumes the
+application.
+
+# Preparing for Checkpoint
+
+As mentioned above, `TR::CompilationInfo::prepareForCheckpoint` handles all the
+coordination needed to prepare the compiler for checkpoint. This is a very
+complex set of operations as it involves coordinating multiple threads, as well
+as being aware of events such as shutdown.
+
+First, this method releases VM Access and acquires the Compilation Monitor. It
+then checks if the checkpoint should be interrupted. It uses the
+`TR::CompilationInfo::_checkpointStatus` member variable to determine the state
+of the checkpoint. The Compilation Monitor is used to guard accesses to the
+checkpoint status; as such, prior to proceeding with the checkpoint process, the
+method needs to check if something like shutdown is currently occurring.
+Suspending threads during shutdown is pointless as the shutdown process will
+wake them up anyway to terminate them. If the checkpoint should be interrupted,
+the thread returns immediately; otherwise, it proceeds to checkpoint[^1].
+
+Next it triggers compilation of any methods that should be compiled proactively
+and waits for them to finish. Then, it calls
+`TR::CompilationInfo::suspendJITThreadsForCheckpoint` which:
+
+1. Suspends Compilation Threads
+2. Suspends the Sampler Thread
+3. Suspends the IProfiler Thread
+
+It then frees any SSL certificates. Finally it sets the Checkpoint Status to
+`TR_CheckpointStatus::READY_FOR_CHECKPOINT_RESTORE`, which indicates that the
+compiler is now ready for checkpoint. Before returning it ensures that it
+releases the Comp Monitor and reacquires VM Access.
+
+As will become evident shortly, all of the subtlety and complexity comes from
+the combination of the facts that  **i.** shutdown, jitdump, or both can occur
+at any point between the start and end of `prepareForCheckpoint`,  **ii.**
+coordinating the checkpointing thread and the Compiler Thread(s) to be suspended
+involves using a monitor different from the one used to coordinate said Compiler
+Thread(s).
+
+## Monitors
+
+**Compilation Monitor:** Referred to below as "Comp Monitor"; the monitor used
+by the compilation infrastructure to coordinate compilation threads, compilation
+entries, etc. (not to be confused with the Compilation Thread Monitor, see
+[Compilation Threads](./CompilationControl.md#compilation-threads) and
+ [Compilation Monitor](./CompilationControl.md#compilation-monitor) for more
+details).
+
+**Compilation Thread Monitor:** Referred to below as "CompThread Monitor"; the
+monitor used by a compilation thread to suspend itself for a checkpoint, or to
+wait for work.
+
+**CR Monitor:** The monitor used by the Checkpointing Thread to wait until the
```

Comment:
`CR Monitor` => `Checkpoint/Restore (CR) Monitor` ? I assume this is what `CR` refers to here

Reply:
It looks like Comp Thread and the Shutdown Thread could also acquire the CR monitor.

Reply:
Yeah they do, but the main purpose of the CR monitor is for the Checkpointing Thread to wait until notified by the others.